wolfcrypt/src/ed25519.c

```diff
@@ -56,30 +56,114 @@ static const byte ed25519Ctx[ED25519CTX_SIZE+1] =
                                              "SigEd25519 no Ed25519 collisions";
 #endif
 
+static int ed25519_hash_init(ed25519_key* key, wc_Sha512 *sha)
+{
+    int ret;
+
+    ret = wc_InitSha512_ex(sha, key->heap,
+#if defined(WOLF_CRYPTO_CB)
+                           key->devId
+#else
+                           INVALID_DEVID
+#endif
+        );
+
+#ifdef WOLFSSL_ED25519_PERSISTENT_SHA
+    if (ret == 0)
+        key->sha_clean_flag = 1;
+#endif
+
+    return ret;
+}
+
+#ifdef WOLFSSL_ED25519_PERSISTENT_SHA
+static int ed25519_hash_reset(ed25519_key* key)
+{
+    int ret;
+    if (key->sha_clean_flag)
+        ret = 0;
+    else {
+        wc_Sha512Free(&key->sha);
+        ret = wc_InitSha512_ex(&key->sha, key->heap,
+#if defined(WOLF_CRYPTO_CB)
+                               key->devId
+#else
+                               INVALID_DEVID
+#endif
+            );
+        if (ret == 0)
+            key->sha_clean_flag = 1;
+    }
+    return ret;
+}
+#endif /* WOLFSSL_ED25519_PERSISTENT_SHA */
+
+static int ed25519_hash_update(ed25519_key* key, wc_Sha512 *sha, const byte* data, word32 len)
+{
+#ifdef WOLFSSL_ED25519_PERSISTENT_SHA
+    if (key->sha_clean_flag)
+        key->sha_clean_flag = 0;
+#else
+    (void)key;
+#endif
+    return wc_Sha512Update(sha, data, len);
+}
+
+static int ed25519_hash_final(ed25519_key* key, wc_Sha512 *sha, byte* hash)
+{
+    int ret = wc_Sha512Final(sha, hash);
+#ifdef WOLFSSL_ED25519_PERSISTENT_SHA
+    if (ret == 0)
+        key->sha_clean_flag = 1;
+#else
+    (void)key;
+#endif
+    return ret;
+}
+
+static void ed25519_hash_free(ed25519_key* key, wc_Sha512 *sha)
```

Comment:
Function only used if `#ifndef WOLFSSL_ED448_PERSISTENT_SHA`. Wrap in this macro check.

Reply:
same story.  moar squinting.